### PR TITLE
Serialize Less Data

### PR DESCRIPTION
**Pull request by jamesmcclain**
This should be have been noticed earlier, but it appears that a lot of the performance issues were/are actually due to serializing too much data.

In this pull request, changes are made to prevent the entire set of (multi-)lines from being serialized and deserialized for every Spark task.  Instead, only those (multi-)lines which are relevant for a specific key are sent along to the task that rasterizes the lines over the associated tiles.   (I am not sure if "task" is the correct Spark-specific nomenclature, but I hope that my meaning is understood.)

~~There is still more that can be done here: the r-tree is not longer required.  Instead, it should be possible to sort the list of lines into appropriate sub-lists directly rather than constructing and querying an r-tree.~~

After a few warm-up runs, this code produces the following result on my work machine
```
{
  "result": {
    "List(42)": 492,
    "List(22)": 15417,
    "List(43)": 1934,
    "List(71)": 412,
    "List(41)": 75479,
    "List(21)": 45149,
    "List(0)": 75,
    "List(24)": 2529,
    "List(31)": 173,
    "List(90)": 27144,
    "List(52)": 11009,
    "List(11)": 3551,
    "List(23)": 8442,
    "List(82)": 12579,
    "List(81)": 18622,
    "List(95)": 709
  }
}curl --data-binary @/tmp/MapshedJob_DRB_VeryLarge.json   0.01s user 0.03s system 0% cpu 4.974 total
```
whereas it could not complete in 10 seconds before.

### CODE REVIEW

**Comment by lossyrob** [comment]
👏 👍

**Comment by jamesmcclain** [comment]
After a little more testing, it appears that constructing and using an r-tree is actually faster than naive filtering, so I think that this can be merge into @rajadain 's branch.

**Comment by rajadain** [comment]
Testing this on my machine now

File: summary/src/main/scala/JobUtils.scala [hunk 1]
```diff
@@ -83,8 +85,7 @@ trait JobUtils {
     * object to read from it.
     *
     * @param   bucket    The name of the S3 bucket
-    * @param   rootPath  The name of the catalog (child of the root directory) in the bucket
-    * @return            An S3LayerReader object
+    * @param   rootPath  The name of the catalog (child of the root directory) in the bucket An S3LayerReader object
```
Comment (rajadain) [inline summary/src/main/scala/JobUtils.scala]: This reads odd

File: summary/src/main/scala/MapshedJob.scala [hunk 2]
```diff
@@ -79,38 +102,61 @@ object MapshedJob extends SparkJob with JobUtils {
...
+          val list = rtree.query(envelope).asScala.map(_.asInstanceOf[MultiLine])
+
+          (k, (extent, list))
+        })
+    val bucketedLinesRDD = sc.parallelize(bucketedLines)
+
+    val joinedRasters = joinRasters(rasterLayers)
+
+    joinedRasters.join(bucketedLinesRDD)
+      .map({ case (key, moop) =>
```
Comment (rajadain) [inline summary/src/main/scala/MapshedJob.scala]: `moop`?

File: summary/src/main/scala/JobUtils.scala (hunk 1 above)
Comment (jamesmcclain) [inline summary/src/main/scala/JobUtils.scala]: That was an accident, will correct.

File: summary/src/main/scala/MapshedJob.scala (hunk 2 above)
Comment (jamesmcclain) [inline summary/src/main/scala/MapshedJob.scala]: Oops

File: summary/src/main/scala/MapshedJob.scala [hunk 3]
```diff
@@ -79,38 +102,61 @@ object MapshedJob extends SparkJob with JobUtils {
...
+    val bucketedLines =
+      rasterLayers
+        .head.map(_._1).collect
+        .map({ k =>
+          val extent = mt(k)
+          val Extent(xmin, ymin, xmax, ymax) = extent
+          val envelope = new Envelope(xmin, xmax, ymin,ymax)
+          val list = rtree.query(envelope).asScala.map(_.asInstanceOf[MultiLine])
+
+          (k, (extent, list))
```
Comment (rajadain) [inline summary/src/main/scala/MapshedJob.scala]: Any reason to prefer nested pairs to a triplet? e.g. `(k, extent, list)`

Comment (rajadain) [inline summary/src/main/scala/MapshedJob.scala]: Ah for the `join`, I see.

File: summary/src/main/scala/MapshedJob.scala [hunk 4]
```diff
@@ -79,38 +102,61 @@ object MapshedJob extends SparkJob with JobUtils {
...
+    joinedRasters.join(bucketedLinesRDD)
+      .map({ case (key, moop) =>
+        val tiles = moop._1
+        val extent = moop._2._1
+        val list = moop._2._2
 
-        // Similarly, we calculate rasterExtent using the first layer's tiles
         val rasterExtent = RasterExtent(extent, tiles.head.cols, tiles.head.rows)
 
         val Extent(xmin, ymin, xmax, ymax) = extent
```
Comment (rajadain) [inline summary/src/main/scala/MapshedJob.scala]: These min and max values are no longer used

**Comment by jamesmcclain** [comment]
I think that all comments prior to this one have been addressed.

**Comment by rajadain** [comment]
In my tests I'm not seeing  a significant speedup 😞 

|           |         | Small | Medium | Large | Very Large |
|-----------|---------|-------|--------|-------|------------|
| *Base*    | 1st Run | 11.75 | 2.4    | 11.08 | 204        |
|           | 2nd Run | 5     | 2.13   | 8.38  | 13.15      |
|           | 3rd Run | 2.92  |        | 39.47 | 8.8        |
|           | 4th Run | 3.75  |        | 10    | 11.5       |
|           |         |       |        |       |            |
| *This PR* | 1st Run | 13    | 3.8    | 13.6  | 12.3       |
|           | 2nd Run | 3.8   | 2.8    | 12.15 | 10.3       |
|           | 3rd Run | 2.7   |        | 13.2  | 55.7       |
|           | 4th Run |       |        |       | 10.93      |

In both cases, I reloaded the worker so as to have a fresh start. Then I ran the *Small* a bunch of times till it stabilized, then ran each one till they stabilized (more or less). Small, Medium, and Large are overlapping, while Very Large doesn't overlap any. The few with extra long times can probably be chalked up to garbage collection / slow internet, but if we exclude the outliers the average is pretty stable for each.

**Comment by jamesmcclain** [comment]
I actually did not see any difference for any except the very large case, but I did measure significant improvement in that case.  I'll measure again and get back to you.

**Comment by jamesmcclain** [comment]
I should also say that I was consistently measuring sub-5 second times this weekend, so I think network traffic is definitely a factor.

**Comment by rajadain** [comment]
Were you testing within the `worker` VM? Or natively on your computer?

**Comment by jamesmcclain** [comment]
Natively.

If I recall, the worker VM has one core.  How many cores does this code have access to in production?

**Comment by rajadain** [comment]
@hectcastro should be able to answer that

**Comment by jamesmcclain** [comment]
For a sufficiently small number of cores, it might make sense to just suppress Spark entirely by doing an early `.collect()` and do the calculation using regular Scala collections.

**Comment by hectcastro** [comment]
For production EC2, each worker has 2 cores.

**Comment by jamesmcclain** [comment]
How much RAM?

**Comment by hectcastro** [comment]
4GB. The workers are `t2.mediums` right now.

**Comment by jamesmcclain** [comment]
I have added a commit with support for sequential processing (eschewing Spark to avoid serialization).  You can try it by using the operationType `RasterLinesJoinSequential` instead of `RasterLinesJoin`.

I have also added a column to my chart above.  Keep in mind that these measurements were taken at different times of day so they are not 100% comparable, but I think that the results in the two right-most column are essentially the same.

@rajadain, I would be interested to see how this does in your testing setup.

**Comment by rajadain** [comment]
Latest timing:

| | | Small | Medium | Large | Very Large |
|---------------------------|---------|--------|------------|--------|--------|
| RasterLinesJoin | 1st Run | 12.245 | 4.084 | 11.803 | 9.051 |
|  | 2nd Run | 4.226 | 3.053 | 11.673 | 7.525 |
|  | 3rd Run | 2.855 | 3.199 | 11.322 | 7.369 |
|  | 4th Run | 3.024 |  |  | 8.358 |
|  |  |  |  |  |  |
| RasterLinesJoinSequential | 1st Run | 10.115 | 2.195 | 7.595 | 8.502 |
|  | 2nd Run | 2.68 | 2.454 | 6.422 | 7.69 |
|  | 3rd Run | 2.482 | 2.375 | 6.866 | 8.172 |
|  | 4th Run | 2.148 |  |  | 12.072 |

I restarted the worker between tests. It does seem like `RasterLinesJoinSequential` is faster overall. I say we should replace `RasterLinesJoin` with it and then merge this in. This is pretty good enough for now, at least until we don't have parallelization effects.

Great work!

**Comment by rajadain** [comment]
+1

**Comment by jamesmcclain** [comment]
It is your choice, but I would be for keeping both the sequential and parallel version and perhaps choosing between them in the frontend.  As implied by the charts above, the particular case in which I see the parallel version being very useful is when there is a large number of tiles but a fairly sparse collection of streams (e.g. using the NHD dataset over a large AoI).

If GitHub had a "drop the mic" icon I would now use it.

**Comment by rajadain** [comment]
Ah, I see now. Very well, let me merge this in.

**Comment by rajadain** [comment]
Re: adding custom memes in a PR, just embed an image:

micdrop

**Comment by jamesmcclain** [comment]
:smile: Okay!

**Pull request merged**